src/java/org/apache/cassandra/schema/Schema.java

```diff
@@ -611,18 +613,42 @@ public SchemaTransformationResult transform(SchemaTransformation transformation,
      * Clear all locally stored schema information and fetch schema from another node.
      * Called by user (via JMX) who wants to get rid of schema disagreement.
      */
-    public synchronized void resetLocalSchema()
```

Comment:
Most general answer to that question is - because we are waiting in this method. The method simply mark a schema that is should be refreshed and wait until it happens. However, the schema would be unable to refresh if we hold a lock because we no changes could be applied then. Note that, the changes are applied from the MigrationCoordinator thread, usually a different to the called thread.